CMakeLists.txt

```diff
@@ -321,7 +321,7 @@ endif()
 
 # Get the current commit ref
 execute_process(
-  COMMAND git describe --tags --always --dirty=-modified
+  COMMAND git describe --tags --always
```

Comment:
So how it is possible to distinguish a patch Mixxx from an unpatched one?

Reply:
I have removed it here, because it is unreliable at this point. The dirty flag is only calculated at configure time. 
If you build a new commit, it is calculated as unmodified. When a source file has been changed, it is not reconfigured. 

I have fixed it by moving the dirty check to cmake/builddate.cmake and let it depend on mixxx-lib. This works reliable except for main.cpp which is OK I think.

Reply:
We could make sure to always run the git info header configuration. We can use https://cmake.org/cmake/help/latest/command/add_custom_target.html for this. What do you think?

Reply:
Idea: We always run the command during every build but only configure a temp file, then compare the content of the present configured file with the temp file. If they don't match, overwrite the configured file with the temp file. That way there should be no unnecessary recompilations and the dirty-flag would work for free, without the complicated hack for detecting the dirty state.

I can take care of that If you want (you're probably busy with investigating the PPA). Just let me know.

As I said I'd also like to split the version info into two parts:
1. Stuff that often changes (i.e. git describe, date) => here we can use the approach I listed above.
2. Stuff that rarely changes (CMAKE_PROJECT_VERSION) => This should be split off into a separate header file that can be included in `defs_urls.h` for example, so that we can replace the hardcoded 2.3 version there on the preprocessor level.

Reply:
What would be improved? 

I am afraid it will not work well and is more hacky than the current solution which uses the normal dependency tree of the underlying build system.

> Idea: We always run the command during every build but only configure a temp file.

I like to have a solution that only kicks in when we create a new binary anyway.

> That way there should be no unnecessary recompilations and the dirty-flag would work for free.

Not really, I guess the current solution is much faster since It already has no unnecessary recompile. 

Splitting up the versions into two files makes only sense if they go into different object files.

This is the case currently. The often changing dirty flag and date are moved into the builddate.cpp.o and the rest into versiondtore.cpp.o.

I don't think that writing the version will be give a notable delay. So we may move everything from version.h into a include free version.cpp.o and put this into an own lib replacing builddate.cpp bx following the same pattern implemented here. 

What do you think?

Reply:
> What would be improved?

`git describe --dirty` would work.

> I am afraid it will not work well and is more hacky than the current solution which uses the normal dependency tree of the underlying build system.

It would use the dependency tree of the build system.
If the values of git describe or the date changed, then we overwrite the configured header file -> Therefore the file time changes and the build system will recompile all files that include it.

It's not really hacky IMHO.

> Not really, I guess the current solution is much faster since It already has no unnecessary recompile.

If no source file changes (i.e. if git describe and date did not change), then the file time will stay the same and the build system shouldn't recompile/relink the mixxx binary.

> Splitting up the versions into two files makes only sense if they go into different object files.

That is the point. I'd like to use CMAKE_PROJECT_VERSION_MAJOR and CMAKE_PROJECT_VERSION_MINOR in `defs_urls.h`. Right now this is not possible without recompiling *every* compilation unit that includes `defs_urls.h` on every single commit.

Reply:
> > What would be improved?
> 
> `git describe --dirty` would work.
> 
> > I am afraid it will not work well and is more hacky than the current solution which uses the normal dependency tree of the underlying build system.
> 
> It would use the dependency tree of the build system.
> If the values of git describe or the date changed, then we overwrite the configured header file -> Therefore the file time changes and the build system will recompile all files that include it.
> 
> It's not really hacky IMHO.
> 
> > Not really, I guess the current solution is much faster since It already has no unnecessary recompile.
> 
> If no source file changes (i.e. if git describe and date did not change), then the file time will stay the same and the build system shouldn't recompile/relink the mixxx binary.
> 
> > Splitting up the versions into two files makes only sense if they go into different object files.
> 
> That is the point. I'd like to use CMAKE_PROJECT_VERSION_MAJOR and CMAKE_PROJECT_VERSION_MINOR in `defs_urls.h`. Right now this is not possible without recompiling *every* compilation unit that includes `defs_urls.h` on every single commit because it's in the same header file as the GIT_DESCRIBE variable.

Reply:
The "-modified" appendix is already in the about box with this PR and that reliable. That was the reason why I have stared this PR. 

The root issue we are facing here, is that we use git describe as a package name and CPack is configured at configure time. 
The system is reconfigured whenever you commit. The first build after commit will produces a new "git describe" string and is almost always not dirty. All following changes to the source do not reconfigure the project. That's why the dirty flag from configure time is unreliable and I have --dirty=-modified in the CMakeList.txt.

For a reliable dirty flag we need to check it whenever a new binary is created, This is done here with an independent git call. The result is appended to the git describe string just like git would do it.

We have now the issue, that the package name does not have the modified appendix with this branch. I consider this not and issue, because it is an edge case and a solution requires a hack to read the modified flag from a file during CPack time, I do avoid to maintain. Is that OK? 

> It would use the dependency tree of the build system.

The penalty would be that the cmake script running git described is executed on every build. The project is reconfigured, whenever the dirty flag changes, not only if the commit hash changes. It is also reconfigured and forces an unnecessary rebuild if a file is changed that would normally not force a rebuild, like skins and controller mappings. 

That the reason why I am not to fond of it. 

> defs_urls.h

We could remove all version strings form defs_urls.h and compose the Urls on the fly using version store. Would it work? 
We may also put getters to VersionStrore and include defs_urls only from there. 
Or do you need the defs_urls for something else?